hist/histv7/inc/ROOT/RHist.hxx

```diff
@@ -300,23 +300,34 @@ using RH3I = RHist<3, int, RHistStatContent>;
 using RH3LL = RHist<3, int64_t, RHistStatContent>;
 ///\}
 
-/// Add two histograms. This is the generic, inefficient version for now; it
-/// assumes no matching axes.
-template <int DIMENSIONS, class PRECISION_TO, class PRECISION_FROM,
+/// Add two histograms.
+///
+/// This operation may currently only be performed if the two histograms have
+/// the same axis configuration, use the same precision, and if `from` records
+/// at least the same statistics as `to` (recording more stats is fine).
+///
+/// In the future, we may either adopt a more relaxed definition of histogram
+/// addition or provide a mechanism to convert from one histogram type to
+/// another. We currently favor the latter path.
+template <int DIMENSIONS, class PRECISION,
           template <int D_, class P_> class... STAT_TO,
           template <int D_, class P_> class... STAT_FROM>
-void Add(RHist<DIMENSIONS, PRECISION_TO, STAT_TO...> &to, const RHist<DIMENSIONS, PRECISION_FROM, STAT_FROM...> &from)
+void Add(RHist<DIMENSIONS, PRECISION, STAT_TO...> &to, const RHist<DIMENSIONS, PRECISION, STAT_FROM...> &from)
 {
-   auto toImpl = to.GetImpl();
-   auto fillFuncTo = toImpl->GetFillFunc();
-   using HistFrom_t = RHist<DIMENSIONS, PRECISION_FROM, STAT_FROM...>;
-   using FromCoord_t = typename HistFrom_t::CoordArray_t;
-   using FromWeight_t = typename HistFrom_t::Weight_t;
-   auto add = [fillFuncTo, toImpl](const FromCoord_t &x, FromWeight_t c) {
-      (toImpl->*fillFuncTo)(x, c);
-      // RODO: something nice with the uncertainty - depending on whether `to` cares
-   };
-   from.GetImpl()->ApplyXC(add);
+   // Enforce "same axis configuration" policy.
+   auto& toImpl = *to.GetImpl();
+   const auto& fromImpl = *from.GetImpl();
+   for (int dim = 0; dim < DIMENSIONS; ++dim) {
+      if (!toImpl.GetAxis(dim).HasSameBinningAs(fromImpl.GetAxis(dim))) {
+         R__ERROR_HERE("HIST") << "Incompatible axis types";
+         // FIXME: Shouldn't this throw an exception or something?
```

Comment:
As the FIXME says, this feels wrong. What is ROOT's policy for erroring our in functions that return void?

- Is it okay to throw an exception in that case?
- Should the function return type be modified to allow returning some sort of error code?
- Is the current "continue as if nothing happened" pattern, which is inspired by code that I saw elsewhere in the codebase, truly the preferred course of action?

Reply:
I'd prefer throwing an exception.

Reply:
Alright, will do that then. Just wanted to check that with you because IIRC ROOT has a CMake flag for disabling exceptions, which suggests that someone once thought that being able to use ROOT with guarantee that no exceptions will be thrown was important.

Reply:
Done.